source/ice_api_private.c

```diff
@@ -0,0 +1,743 @@
+    }
+    else
+    {
+        controllingAgentCandidatePriority = localCandidatePriority;
+        controlledAgentCandidatePriority = remoteCandidatePriority;
+    }
+
+    /* https://tools.ietf.org/html/rfc5245#appendix-B.5 */
+    candidatePairPriority = ( ( ( uint64_t ) 1 << 32 ) *
+                              ( controllingAgentCandidatePriority >= controlledAgentCandidatePriority ?
+                                controlledAgentCandidatePriority : controllingAgentCandidatePriority ) +
+                              ( 2 * ( controllingAgentCandidatePriority >= controlledAgentCandidatePriority ?
+                                      controllingAgentCandidatePriority : controlledAgentCandidatePriority ) ) +
+                              ( controllingAgentCandidatePriority > controlledAgentCandidatePriority ? 1 : 0 ) );
+
+    return candidatePairPriority;
+}
+
+/*----------------------------------------------------------------------------*/
+
+IceResult_t Ice_FinalizeStunPacket( IceContext_t * pContext,
+                                    StunContext_t * pStunCtx,
+                                    const uint8_t * pPassword,
+                                    size_t passwordLength,
+                                    size_t * pStunMessageBufferLength )
+{
+    IceResult_t iceResult = ICE_RESULT_OK;
+    StunResult_t stunResult = STUN_RESULT_OK;
+    uint8_t messageIntegrity[ STUN_HMAC_VALUE_LENGTH ];
+    size_t messageIntegrityLength = STUN_HMAC_VALUE_LENGTH;
+    uint8_t * pIntegrityCalculationData = NULL;
+    uint16_t integrityCalculationDataLength = 0;
+    uint8_t * pFingerprintCalculationData = NULL;
+    uint16_t fingerprintCalculationDataLength = 0;
+    uint32_t messageFingerprint = 0;
+
+    if( pPassword != NULL )
+    {
+        /* Add Integrity attribute. */
+        stunResult = StunSerializer_GetIntegrityBuffer( pStunCtx,
+                                                        &( pIntegrityCalculationData ),
+                                                        &( integrityCalculationDataLength ) );
+
+        if( stunResult == STUN_RESULT_OK )
+        {
+            iceResult = pContext->cryptoFunctions.hmacFxn( pPassword,
+                                                           passwordLength,
+                                                           pIntegrityCalculationData,
+                                                           ( size_t ) integrityCalculationDataLength,
+                                                           &( messageIntegrity[ 0 ] ),
+                                                           &( messageIntegrityLength ) );
+
+            if( iceResult == ICE_RESULT_OK )
+            {
+                stunResult = StunSerializer_AddAttributeIntegrity( pStunCtx,
+                                                                   &( messageIntegrity[ 0 ] ),
+                                                                   ( uint16_t ) messageIntegrityLength );
```

Comment:
Possible data loss casting `size_t` (which could be 64 bits) to `uint16_t`

Reply:
Thanks. I have updated the code.